#' MmapprParam Class and Constructor
#' 
#' \code{MmapprParam} stores parameters for running \code{\link{mmappr}}.
#'
#' @slot wtFiles Character vector,
#'   \code{\link[Rsamtools]{BamFile}}, or
#'   \code{\link[Rsamtools]{BamFileList}} containing
#'   BAM files for the wild-type pool to be analyzed.
#' @slot mutFiles Character vector,
#'   \code{\link[Rsamtools]{BamFile}}, or
#'   \code{\link[Rsamtools]{BamFileList}} containing
#'   BAM files for the mutant pool to be analyzed.
#' @slot refFasta The path to the fasta file genome.
#' @slot gtf The path to a gtf-formatted annotation file. 
#' @slot outputFolder Length-one character vector specifying where to save
#'   output, including a \code{\linkS4class{MmapprData}} stored as
#'   \code{mmappr_data.RDS}, \code{mmappr2.log}, a \code{.tsv} file
#'   for each peak chromosome containing candidate mutations, and PDF plots
#'   of both the entire genome and peak chromosomes. Defaults to an
#'   automatically generated \code{mmappr2_<timestamp>}.
#' @slot includeScaffolds Logical indicating whether non-standard chromosomes
#'   should be included. False (default) trims the seqnames of the reference to 
#'   include only standard chromosome names based on UCSC or Ensembl 
#'   conventions. In all cases, the mitochondrial chromosome is removed.
#' @slot minDepth Length-one integer vector determining minimum depth
#'   required for a position to
#'   be considered in the analysis. Defaults to 20.
#' @slot homozygoteCutoff Length-one numeric vector between \code{0} and
#'   \code{1} specifying threshold for throwing out base pairs on account
#'   of homozygosity. Positions with high major allele frequency in the
#'   wild-type pool are unlikely to exhibit polymorphism and are thus thrown
#'   out when they exceed this cutoff. Defaults to \code{0.95}.
#' @slot minBaseQuality Length-one numeric vector indicating minimum base
#'   call quality to consider in analysis. Read positions with qualities
#'   below this score will be thrown out. Defaults to 20.
#' @slot minMapQuality Length-one numeric vector indicating minimum read
#'   mapping quality to consider in analysis. Reads with qualities below
#'   this score will be thrown out. Defaults to 30.
#' @slot fileAggregation A length-one character vector determining strategy
#'   for aggregating base calls when multiple wild-type or multiple mutant
#'   files are provided.
#'   When 'sum', average base call proportions are computed using
#'   the read counts from each file, effectively weighting files
#'   with higher counts at a given position. When equal to 'mean', the
#'   base call proportions as well as read depths, rather than the absolute count,
#'   are averaged across files, which is useful when you want to weight each
#'   replicate evenly without regards to differing depth.
#' @slot distancePower Length-one numeric vector determing to what power
#'   Euclidean distance values are raised before fitting. Higher powers tend
#'   to increase high values and decrease low values, exaggerating the
#'   variation in the data. Default of 4.
#' @slot peakIntervalWidth Length-one numeric vector between \code{0} and
#'   \code{1} specifying desired width of linkage region(s). The default value
#'   of \code{0.95}, for example, yields peak regions defined as including the
#'   top 95\% of SNPs in the peak region, as determined by the peak
#'   resampling distribution.
#' @slot loessOptResolution Length-one numeric vector between \code{0} and
#'   \code{1} specifying
#'   desired resolution for Loess fit optimization. The default of \code{0.001},
#'   for example, indicates that the span ultimately chosen will perform better
#'   than its neighbor values at \code{+-0.001}.
#' @slot loessOptCutFactor Length-one numeric vector between \code{0} and
#'   \code{1} specifying how aggressively the Loess
#'   optimization algorithm proceeds. For example, with a default of \code{0.1}
#'   different spans at intervals of \code{0.001} would be evaluated after
#'   intervals of \code{0.01}.
#' @slot vepFlags Optional \code{\link[ensemblVEP]{VEPFlags}}
#'   object containing runtime options for Ensembl's Variant Effect Predictor.
#'   See vignette for details. Generated by default.
#' @slot vepImpact A number from 1 to 4 designating the mutation impact ratings  
#'   to be included. 1 = Hight, 2 = High and Moderate, 3 = High, Moderate, and 
#'   Low, and 4 = High, Moderate, Low, and Modifier.
#' @slot refGenome A \code{\link[GmapGenome]{GmapGenome}} object. This is 
#'   generated internally from the provided refFasta file to ensure 
#'   consistency.
#' @rdname MmapprParam
#' @export

setClass("MmapprParam",
         representation(
           wtFiles = "BamFileList", 
           mutFiles = "BamFileList",
           refFasta = "character",
           refGenome = "GmapGenome",
           gtf = "character", 
           outputFolder = "character", 
           includeScaffolds = "logical",
           minDepth = "numeric",
           homozygoteCutoff = "numeric", 
           minBaseQuality = "numeric",
           minMapQuality = "numeric",
           fileAggregation = "character", 
           distancePower = "numeric",
           peakIntervalWidth = "numeric", 
           loessOptResolution = "numeric",
           loessOptCutFactor = "numeric", 
           vepFlags = "VEPFlags",
           vepImpact = "numeric"))


#' MmapprData Class
#' 
#' Stores data from each step of the MMAPPR2 pipeline.
#'
#' @slot param \code{\linkS4class{MmapprParam}} object storing parameters
#'   used in analysis.
#' @slot snpDistance List containing raw counts and Euclidean distance data for
#'   each chromosome. After \code{\link{calculateDistance}}, chromosomes with
#'   sufficient data should have \code{$wtCounts},
#'   \code{$mutCounts}, and \code{$distanceDf} populated. After
#'   \code{\link{loessFit}}, the \code{$distanceDf} element for each chromosome
#'   list is replaced with a \code{$loess} element.
#' @slot peaks List of chromosomes containing peak regions. Initialized after
#'   \code{\link{prePeak}} and populated with density function after
#'   \code{\link{peakRefinement}}.
#' @slot candidates List containing \code{\link[GenomicRanges]{GRanges}} objects
#'   with snps, VEP predicted effects, and differential expression data
#'   for each peak, all resulting from \code{\link{generateCandidates}} 
#'   function.
#'
#' @aliases MmapprData
#' @export
#' @seealso \code{\link{mmappr}}, \link{MmapprData-getters}

setClass("MmapprData",
         representation(
             param="MmapprParam",
             snpDistance="list",
             peaks="list",
             candidates="list"
         )
)

